src/dispatch/experimental/durable/frame.c

```diff
@@ -120,44 +280,245 @@ static const char *get_type_name(PyObject *obj) {
+    // If the object isn't a PyGenObject, PyCoroObject or PyAsyncGenObject, it may
+    // still be a coroutine, for example a PyCoroWrapper. CPython unfortunately does
+    // not export a function that checks whether a PyObject is a PyCoroWrapper. We
+    // need to check the type name string.
     const char *type_name = get_type_name(obj);
     if (!type_name) {
         return NULL;
     }
     if (strcmp(type_name, "coroutine_wrapper") == 0) {
+        // FIXME: improve safety here, e.g. by checking that the obj type matches a known size
         PyCoroWrapper *wrapper = (PyCoroWrapper *)obj;
+        // Cast the inner PyCoroObject to a PyGenObject. See the comment above.
         return (PyGenObject *)wrapper->cw_coroutine;
     }
     PyErr_SetString(PyExc_TypeError, "Input object is not a generator or coroutine");
     return NULL;
 }
 
-static InterpreterFrame *get_interpreter_frame(PyGenObject *gen_like) {
-    struct _PyInterpreterFrame *frame = (struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+static Frame *get_frame(PyGenObject *gen_like) {
+#if PY_MINOR_VERSION == 10
+    Frame *frame = (Frame *)(gen_like->gi_frame);
+#elif PY_MINOR_VERSION == 11
+    Frame *frame = (Frame *)(struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+#elif PY_MINOR_VERSION == 12
+    Frame *frame = (Frame *)(struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+#elif PY_MINOR_VERSION == 13
+    Frame *frame = (Frame *)(struct _PyInterpreterFrame *)(gen_like->gi_iframe);
+#endif
     assert(frame);
-    return (InterpreterFrame *)frame;
+    return frame;
+}
+
+static PyCodeObject *get_frame_code(Frame *frame) {
+#if PY_MINOR_VERSION == 10
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 11
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 12
+    PyCodeObject *code = frame->f_code;
+#elif PY_MINOR_VERSION == 13
+    PyCodeObject *code = (PyCodeObject *)frame->f_executable;
+#endif
+    assert(code);
+    return code;
+}
+
+static int get_frame_lasti(Frame *frame) {
+#if PY_MINOR_VERSION == 10
+    return frame->f_lasti;
+#elif PY_MINOR_VERSION == 11
+// https://github.com/python/cpython/blob/3.11/Include/internal/pycore_frame.h#L69
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->prev_instr);
+    return (int)((intptr_t)frame->prev_instr - (intptr_t)_PyCode_CODE(code));
+#elif PY_MINOR_VERSION == 12
+// https://github.com/python/cpython/blob/3.12/Include/internal/pycore_frame.h#L77
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->prev_instr);
+    return (int)((intptr_t)frame->prev_instr - (intptr_t)_PyCode_CODE(code));
+#elif PY_MINOR_VERSION == 13
+// https://github.com/python/cpython/blob/v3.13.0a5/Include/internal/pycore_frame.h#L73
+    PyCodeObject *code = get_frame_code(frame);
+    assert(frame->instr_ptr);
+    return (int)((intptr_t)frame->instr_ptr - (intptr_t)_PyCode_CODE(code));
+#endif
+}
+
+void set_frame_lasti(Frame *frame, int lasti) {
```

Comment:
Should this function be static for consistency with the other functions in this file?